### DISCUSSION TITLE
How to get logs for pyright-langserver crash?

### DISCUSSION THREAD

**Discussion by damca** [microsoft/pyright#5917] [General]
The `pyright` server connects for a few seconds, provides accurate intellisense and then dies. I would like to troubleshoot the crash, but don't know how to find the logs. My overall goal is to get a `pyright` server within a docker container to communicate with Neovim on the host. 

Setup:
* Running `docker exec -i conda /bin/bash --login -c 'pyright-langserver --stdio'` from host starts a usual server. Note that if `-it` is passed then the Neovim side crashes with the error `the input device is not a TTY` (as found in `~/.local/state/nvim/lsp.log`).
* I made an executable file with the above line called `docker_conda_pyright`. And start the language server with the following lua function:

```lua
start_docker_pyright = function()
  local client = vim.lsp.start_client {
    cmd = { [[docker_conda_pyright]] },
  }
  local bufnr = vim.api.nvim_get_current_buf()
  vim.lsp.buf_attach_client(bufnr, client)
end
```

When calling `start_docker_pyright` on a file, the language server seems to work for ~5 seconds and then crashes. Any thoughts on how to best proceed troubleshooting this?

**Comment by jlucier** [comment]
@damca did you even get this working? I've been trying off and on to accomplish the same thing to no avail. It's as if it processes the buffer and then "shuts down" the server.

**Comment by rchiodo** [comment]
If Pyright is crashing, you can run it from the command line usually to reproduce the crash. Then you'd be able to just debug the pyright code.

**Reply by jlucier** [reply]
I'm not seeing anything of interest coming through on `stderr`, which is really the only thing we've got access to since `stdio` is being used for LSP communication with neovim.

I'm not actually convinced it's crashing, it seems like maybe it's just exiting as soon as it finishing processing the buffer neovim is requesting that it process. Is it possible that `pyright-langserver`, when run this way, might act that way? Or must I be doing something wrong?

**Reply by rchiodo** [reply]
If you start Pyright as a language server (meaning the server bundle), it should stay running. The only thing that would close it would be a crash or sending a [shutdown ](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#shutdown) message.

**Comment by Mazyod** [comment]
I've made a really silly mistake of setting the `processId` to `1` for some reason. Setting it to `None` made everything work well.

I reached the point of cloning Pyright locally, peppering it with logs, and still there were no logs indicating the issue. Perhaps it's an OS thing or logs are being misdirected, idk.

-------

I've been experiencing the exact issue. I have a Python wrapper around the `pyright-langserver --stdio` process and communicating through subprocess standard I/O.

### Things I've tried:
- In the tests I am writing, if I add a sleep of 5 seconds, then attempt to send a request, I get an error that the connection was refused and/or was reset.
- I tried using a for-loop that writes every second, and it still disconnects after the 5 second mark is up, so it's safe to say this isn't some sort of idle timeout.
- Using the exact same code, just replacing `pyright-langserver` with `jedi-language-server`, everything works fine.
- Use `$(npm root -g)/pyright/langserver.index.js --stdio` as recommended [here](https://github.com/microsoft/pyright/issues/4446), but got the same result.

The fact that I am experiencing the same issue from Python seems telling that the issue might be more prevalent.

**Comment by jy-lefort** [comment]
I had the same problem, which I've debugged using strace in the container:

```
kill(1313, 0)                           = -1 ESRCH (No such process)
```

After that log line, pyright exits. This is because `kill(PID, 0)` (which pyright uses to check if the lsp client is still running) fails in the container because it uses a PID namespace.

Fix for Neovim: https://github.com/neovim/neovim/issues/14504

Fix for Emacs:

```
(advice-add 'lsp-request-async :filter-args
            (lambda (args)
              (when (eq major-mode 'python-mode)
                (let ((method (nth 0 args)))
                  (when (string= method "initialize")
                    (let ((params (nth 1 args)))
                      (plist-put params :processId nil)))))
              args))
```
